File: C++/RC Mecanum Bot/src/Tasks/TaskLowBatteryDetect.cpp
#include <Tasks/TaskLowBatteryDetect.h>
#include "Platform/Hardware.h"

TaskLowBatteryDetect::~TaskLowBatteryDetect(){}
bool TaskLowBatteryDetect::GetBatteryIsLow(){
	return BatteryIsLow;
}
void TaskLowBatteryDetect::SetBatteryIsLow(bool state){
	BatteryIsLow = state;
}
/* ILoopable */
void TaskLowBatteryDetect::OnStart() {
}
void TaskLowBatteryDetect::OnStop() {
}
bool TaskLowBatteryDetect::IsDone() {
	return false;
}
void TaskLowBatteryDetect::OnLoop() {
	float vbat;

	/* get the average voltage from a couple talons */
	vbat = 0;
	vbat += Hardware::leftFront->GetBusVoltage();
	vbat += Hardware::rightFront->GetBusVoltage();
	vbat *= 0.5f;

	if (vbat > 10.50) {
		_downCount = 0;
		if (_upCount < 100)
			++_upCount;
	} else if (vbat < 10.00) {
		_upCount = 0;
		if (_downCount < 100)
			++_downCount;
	}

	if (_downCount > 50) {
		BatteryIsLow = true;
	} else if (_upCount > 50) {
		BatteryIsLow = false;
	} else {
		//don't change filter ouput
	}
}

File: HERO C#/RC Mecanum Bot/Tasks/TaskLowBatteryDetect.cs
/**
 * Task tracking the robot battery voltage.  If the battery voltage is considered low,
 * other subsystem tasks will cut the motor speed to signal user to change the battery.
 */

namespace HERO_Mecanum_Drive_Example
{
    public class TaskLowBatteryDetect : CTRE.Phoenix.Tasking.ILoopable
    {
        int _dnCnt = 0;
        int _upCnt = 0;

        public bool BatteryIsLow { get; private set;}
      
        /* ILoopable */
        public bool IsDone()
        {
            return false;
        }

        public void OnLoop()
        {
            float vbat;

            /* get the average voltage from a couple talons */
            vbat = 0;
            vbat += Platform.Hardware.leftFrnt.GetBusVoltage();
            vbat += Platform.Hardware.rghtFrnt.GetBusVoltage();
            vbat *= 0.5f;

            if (vbat > 10.50)
            {
                _dnCnt = 0;
                if (_upCnt < 100)
                    ++_upCnt;
            }
            else if (vbat < 10.00)
            {
                _upCnt = 0;
                if (_dnCnt < 100)
                    ++_dnCnt;
            }

            if (_dnCnt > 50)
            {
                BatteryIsLow = true;
            }
            else if (_upCnt > 50)
            {
                BatteryIsLow = false;
            }
            else
            {
                //don't change filter ouput
            }
        }

        public void OnStart()
        {
        }

        public void OnStop()
        {
        }
    }
}
